src/forms/mod.rs

```diff
@@ -0,0 +1,159 @@
+//! Forms within your terminal
+use std::{any::Any, collections::HashMap};
+
+use crate::{events::Event, rect_style::BorderStyle, screen::Screen};
+
+mod choices;
+pub mod constraints;
+mod form;
+mod text;
+
+pub use choices::Checkbox;
+pub use choices::Radio;
+use crossterm::style::Color;
+pub use form::Form;
+pub use text::HiddenText;
+pub use text::Text;
+
+use self::constraints::FormConstraint;
+
+/// Helper trait to allow downcasting FormFields
+pub trait ToAny {
+    fn to_any(self) -> Box<dyn Any>;
+}
+impl<T: 'static> ToAny for T {
+    fn to_any(self) -> Box<dyn Any> {
+        Box::new(self)
+    }
+}
+
+/// Necessary functions to build a Form Field
+pub trait FormField: ToAny {
+    /// Base function to allow building the Field programmatically (e.g. in Forms)
+    ///
+    /// This function is separate to `new` to allow Fields to define custom constructors
+    fn make(w: u32, options: FormOptions) -> Self
+    where
+        Self: Sized;
+
+    /// Reset the field
+    fn reset(&mut self);
+
+    /// Get the width of the field
+    fn get_width(&self) -> u32;
+    /// Get the height of the field
+    fn get_height(&self) -> u32;
+    fn get_min_height(&self) -> u32 {
+        1
+    }
+    /// Resize (if possible) the field
+    fn resize(&mut self, w: u32, h: u32);
+
+    /// This function is the heart of FormFields : it allows the form to handle itself by passing a ConsoleEngine Event to it.
+    fn handle_event(&mut self, event: Event);
+
+    /// Set the active state of a field (if applicable)
+    fn set_active(&mut self, active: bool);
+    /// Checks if the state of a field is active
+    fn is_active(&self) -> bool;
+
+    /// Allow the field to validate its content by itself.
+    ///
+    /// Make sure to call [self_validate](#method.self_validate) in it,
+    /// so the Validation Constraints provided by FormOptions will be automatically checked against the output of the field
+    fn validate(&self, validation_result: &mut FormValidationResult);
+    /// Get the output of the field
+    fn get_output(&self) -> FormValue;
```

Comment:
I disagree, because Form also implement `get_output`, giving a HashMap of each of his fields.
`get_result_unvalidated`, on the other hand, returns an `Option<FormOutput>` that depends on the presence of a given field, also `get_result_unvalidated` takes an argument while `get_output` does not.

Reply:
To me the name seems too different, since `get_result_unvalidated` is just a map to field's `get_output`. Maybe rename `get_result_unvalidated` to `get_field_output` or `get_output_of_field`?

Reply:
This might do the job, I named it that way due to its close relation with the get_result (it's more or less an error-free version of it)

Reply:
I'm not sure which one would be better now. Maybe also rename `get_result` to `get_validated_field_output`?

Reply:
I prefer your suggestion on that, we keep the output name everywhere so no confusion about what we will get.
Tell me if it looks better for you

Reply:
Yep, this looks better